README.md
# #D (hashD)
A fully decentralized authenticated broadcast/discovery system based on gossip and proof of work

Original Paper https://hashd.in/hashd-in-draft0/


## PoC - ethDenver

- [ ] PoW hashing Alorithm SHA256 - moving to coocko cycle
- [ ] Merkle Tree hash SHA256 - moving to blake2b or keccak
- [ ] Signatures ECDSA secp256k1 


### CLI API

Install with pip using `pip install --user .`
Prefix commands with `python` using the module flag, i.e., `python -m hashd init`

- [ ] hashd init - implemented
  - Create new block with prevBlk = 0
  - Create two new keys
    - hot key
    - cold key output as a mnemonic
- [ ] hashd addWork blkSig - adds proof of work to a block signature - implemented
- [ ] hashd hashBench - benchmark and show lowest hash after 1,10,100 seconds - implemented
- [ ] hashd set -v 10 "Message" "tag1,tag2" image.jpg - partially implemented
- [ ] hashd get -v 10 type="value"
- [ ] hashd config threshold (forward|store-headers|store-broadcast|store-data)
- [ ] hashd verify KEY - block info
- [ ] hashd explore - p2p search


### Message Types
| message | Contents | Size | Compute| Forward Rules | description|
|---------|----------|------|--------|---------------|------------|
|nonceRelay  | BlkSig + nonce | 80b | 1 hash| does the hash verify & meet threshold| Relay additional PoW|
|headerRelay | nonceRelay + header| 80b+100b| 1 signature| does signature verify| Relay new blockheaders|
|bCastRelay  | headerRelay + hash[A:B] + type + value| 180b+64b+128b| 4 hashes| do all the hashes verify | Add new k,v to database|
|authRelay   | bCastRelay + Auth Script + hashes| 372b+ script size| more hashes and script validation| do all the hashes verify & script is well formed| set new consensus rules for chain|
|dataRelay   | bCastRelay + data + hashes |372b+merkle tree & leaves| n hashes for n data verfification |verify hashes|  Store data of another node's block|

```                                                                            Bytes        Name
                  PoW Hash                      | 1 comparison             | 32 |    |    | Broadcast Hash
                     |                          |                          | sig|nonc|    |
        Block Signature || nonce                | 1 hash verification      | 32 | 16 |    |
               |                                |                          |root|Psig|Pkey|
(Op(4b)||Merkle Root||PrevBlkSig)sign+(pub-key) | 1 signature verification | 32 | 32 | 64 | Blockheader + 4bytes for bit tag field
            /     \                             | 1 hash verification      |    |    |    |
           /       \                            |                          |    |    |    |
          /         \                           |                          |    |    |    |
         /           \                          |                          |    |    |    |
      hashA         hashB(root of block data)   | 1 hash verification(A)   | 32 | 32 |    | hashA tree - Broadcast Tree
      /   \         /   \                       |                          |    |    |    | hashB tree - Data Tree
     /     \                                    |                          |    |    |    |
    /       \                                   |                          |    |    |    |
 hashC     hashD                                | 2 hash verifications(C,D)| 32 | 32 |    |
   |         |                                  |                          |    |    |    |
 type      value                                | 2 data fields            | 64 | 64 |    |
```

| OP Value | Type | Value | Description|
|----------|-----|-------|------------|
| 0        |No broadcast| 0|0|NA|
| 1        |Key Replacement| [current key, hash(cold key)] | Declares the new replacement key and the current key|
| 2        |Well Known| String| This would be a name the chain/id want's to be know by|
| 3        |Service Endpoint| JSON object | containing protocol and address tuples of supported endpoints |  
| 4        |Schema| JSON {opVal:Ttype}| the last schema for a chain supercedes eariler ones on the same opval |
| 5        |Tweet| [hashtag1,hashtag2...]| Simple tweet data type, message data is in the data side of the merkle tree|


NEWS.txt
News
====

0.1-dev
---

*Release date: 15-Mar-2019*

* Placeholder for initial release


componentFunctions.py
import hashlib
import time
import binascii
from mnemonic import Mnemonic
import pickledb
import socket
from ecdsa import SigningKey, VerifyingKey, SECP256k1
db = pickledb.load('test.db',True)
m = Mnemonic('english')

#HOST = ''
#PORT = 44444
#s = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
#s.bind((HOST,PORT))
#s.listen(1)
#conn, addr = s.accept()

def get_ip():
    s = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
    try:
        # doesn't even have to be reachable
        s.connect(('10.255.255.255', 1))
        IP = s.getsockname()[0]
    except:
        IP = '127.0.0.1'
    finally:
        s.close()
    return IP

def bCastMerkle(opType,value):
    hsh = hashlib.sha256()
    hsh.update(opType)
    hashC = hsh.digest()
    hsh = hashlib.sha256()
    hsh.update(value)
    hashD = hsh.digest()
    hsh = hashlib.sha256()
    hsh.update(hashC)
    hsh.update(hashD)
    hashA = hsh.digest()
    return [hashC,hashD,hashA]


def mRoot(hashA,hashB):
    hsh = hashlib.sha256()
    hsh.update(hashA)
    hsh.update(hashB)
    root = hsh.digest()
    return root

def sliceOp(op): #slices integer value into a 4byte array to ensure correctness when constructing the block
    opB = bytearray(4)
    for i in range(4):
        opB[i] = (op >> (24 - i * 8) & 255)
    return opB

def makeBlock(op,mRoot,prevBlkSig,sk):
        opPad = sliceOp(op)
        sigField = opPad + mRoot + prevBlkSig
        #print(sigField)
        blkSig = sk.sign(sigField)
        return blkSig

def addWork(blockSig, duration):
    nonce = 0
    low = bytearray(32)
    for i in range(32):
        low[i] = 255
    start = time.time()
    now = start
    while (start + duration) > now: 
        now = time.time()
        h = hashlib.sha256()
        h.update(blockSig)
        h.update(nonce.to_bytes(16,'little',signed=False))
        crnt = h.digest()
        if crnt < low:
            low = crnt
            lowNonce = nonce
        nonce = nonce + 1 
    lowout = binascii.hexlify(low)
    print(lowout)
    print(lowNonce)
    return lowNonce



def init():
    dsk = SigningKey.generate(curve=SECP256k1)
    dvk = dsk.get_verifying_key()
    isk = SigningKey.generate(curve=SECP256k1)
    ivk = isk.get_verifying_key()
    h = hashlib.sha256()
    h.update(ivk.to_string())
    ikHash = h.digest()
    print("Write these words down, this will allow you to recover your identity if this device is hacked or stolen(BIP39):")
    print("================================================")
    print(m.to_mnemonic(isk.to_string())) #replace with mnemonic!!!
    print("================================================")
    bCast = bCastMerkle(b"Key Replacement",dvk.to_string()+ikHash)
    mr = mRoot(bCast[2], bytearray(32))
    blkSig = makeBlock(1,mr,bytearray(64),dsk)
    return blkSig, dsk

    
def say(op,typ, value,prevBlk, dataHash, sk,workDur):
    lhashes = bCastMerkle(typ,value)
    mr = mRoot(lhashes[2],dataHash)
    blkSig = makeBlock(op,mr,prevBlk,sk)
    nonce = addWork(blkSig, workDur)
    return blkSig
    
def tweet(prevBlk, tweet, hashTag, sk, workDur):
    h = hashlib.sha256()
    h.update(tweet.encode('utf-8'))
    dataHash = h.digest()
    tweet.encode('utf-8')
    return say(5,b"Tweet",hashTag.encode('utf-8'),prevBlk,dataHash,sk,workDur)
    


# ## Initialize a New Identity
# 1. Generates two keys
#     - One key is stored on the device(Hot)
#     - Second key is out put as a BIP39 Mnemonic(Cold)
# 2. The hash of the Cold key's public key is included in the first block of this Identity
# 3. The merkleroot and previous block signature(all zeros since this the first block) are signed by the Hot key
     

blkSig0, deviceSK = init()


# ## Broadcast a Well Known name for this idchain
# This example does 10 seconds of proof of work on the second block in our chain to broadcast a name we want to be known as
# - we can add more work to this later 
# - we have the option to have multiple names


blkSig1 = say(2,b"Well Known",b"HashD",blkSig0,bytearray(32),deviceSK,10)


# ## Broadcast a Service Endpoint
# Now that people know our name we want then to be able to connect to us
# 
# Since we have already 10 seconds of PoW we don't need to add much work to broadcast a service endpoint
# 
# Metcalf's Law means adding more peers is a postive sum game, other nodes will value the data that allows them to connect to more peers and information. Self incentivizing them to keep more of this message type for a given amount of PoW


ip = get_ip()

blkSig2 = say(3,b"Service Endpoint",ip.encode('utf-8'),blkSig1,bytearray(32),deviceSK,2)


# ## Broadcast a Tweet 
# Because everyone wants to be heard we have to shout very loudly, or turn up the transmit power of our virtual radio broadcast
# 
# We will do 60 seconds of hash, which on this laptop factoring in depreciation and energy will cost me about $0.0002, which means I could buy 800KB on a modern NVME SSD
# 
# To store the data that would allow people to discover the tweet a node would only need to store 378b verify a few hashes and a signature
# 
# The Nash Equilibrium of cost to broadcast vs cost to network for storage means roughly 2100 nodes could store this broadcast with even cost to publisher and listeners


blkSig3 = tweet(blkSig2,"First","ethDenver",deviceSK,60)

